src/sizes/retained_size_by_classloaders.cpp

```diff
@@ -0,0 +1,102 @@
+// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+
+#include <vector>
+#include "retained_size_by_classloaders.h"
+#include <algorithm>
+#include <numeric>
+
+
+jvmtiError RetainedSizeByClassLoadersAction::tagObjectsByClassLoader(jclass *classes, jint* cnt, jobject classLoader, jsize classLoaderIndex) {
+    jvmtiError err = JVMTI_ERROR_NONE;
+    std::vector<jclass> filteredClasses;
+    for (jsize i = 0; i < *cnt; i++) {
+        jobject rootClassLoader = getClassLoader(env, classes[i]);
+        if (!env->IsSameObject(rootClassLoader, NULL)) {
+            if (isEqual(env, classLoader, rootClassLoader)) {
+                filteredClasses.emplace_back(classes[i]);
+            }
+        }
+    }
+
+    err = tagObjectsOfClassLoaderClasses(toJavaArray(env, filteredClasses), classLoaderIndex);
+    if (!isOk(err)) return err;
+    if (shouldStopExecution()) return MEMORY_AGENT_INTERRUPTED_ERROR;
+    return err;
+}
+
+jlong RetainedSizeByClassLoadersAction::tagOtherObjects(jclass *classes, jint* cnt, jobjectArray classLoadersArray) {
+    jvmtiError err = JVMTI_ERROR_NONE;
+    std::vector<jclass> filteredClasses;
+    for (jsize i = 0; i < *cnt; i++) {
+        jobject rootClassLoader = getClassLoader(env, classes[i]);
+        if (!env->IsSameObject(rootClassLoader, NULL)) {
+            for (jsize j = 0; j < env->GetArrayLength(classLoadersArray); j++) {
+                if (isEqual(env, env->GetObjectArrayElement(classLoadersArray, j), rootClassLoader)) {
+                    filteredClasses.emplace_back(classes[i]);
+                    break;
+                }
+            }
+        }
+    }
+
+    std::vector<jclass> nonfilteredClasses;
+    for (jsize i = 0; i < *cnt; i++) {
+        if (std::find(filteredClasses.begin(), filteredClasses.end(), classes[i]) == filteredClasses.end()) {
```

Comment:
It would be much better to store filtered classes in `std::unordered_set`. This way the time complexity of this loop would be O(n).